HW0/P4/P4.py

```diff
@@ -23,54 +23,148 @@
+    ppts_trans = np.transpose(ppts)
+    #####################
+     
+    x_coords = ppts_trans[:,0]
+    y_coords = ppts_trans[:,1]
+    z_coords = ppts_trans[:,2]
+    ax.plot(x_coords, y_coords, z_coords,
+            '.g', label='Observed trajectory')
 
     #####################
     # Part 3:
     # Use the initial conditions and propagation matrix for prediction
     #####################
 
-    # A = ?
-    # a = ?
-    # s = ?
-
+    A = np.array([ [1,0,0,dt,0,0],
+                   [0,1,0,0,dt,0],
+                   [0,0,1,0,0,dt],
+                   [0,0,0,1-c*dt,0,0],
+                   [0,0,0,0,1-c*dt,0],
+                   [0,0,0,0,0,1-c*dt] ])
+    a = np.array([ [0,0,0,0,0,g*dt] ]).T
+    s0 = np.array([ [0,0,2,15,3.5,4.0] ]).T
+    
     # Initial conditions for s0
     # Compute the rest of sk using Eq (1)
+    sk = s0
+    s  = s0
+
+    i = 1
+    while i < K:
 
-    # ax.plot(x_coords, y_coords, z_coords,
-    #         '-k', label='Blind trajectory')
+       sk = np.dot(A, sk) + a
+       s  = np.c_[s, sk]
+  
+       i += 1
+
+    ###### print s.shape  
+    x_coords = s[0,:] 
+    y_coords = s[1,:]
+    z_coords = s[2,:]
+    ax.plot(x_coords, y_coords, z_coords,
+            '-k', label='Blind trajectory')
 
     #####################
     # Part 4:
     # Use the Kalman filter for prediction
     #####################
 
-    # B = ?
-    # C = ?
+    B = np.array([ [bx,0,0,0,0,0],
+                   [0,by,0,0,0,0],
+                   [0,0,bz,0,0,0],
+                   [0,0,0,bvx,0,0],
+                   [0,0,0,0,bvy,0],
+                   [0,0,0,0,0,bvz] ])
+    
+    C = np.array([ [rx,0,0,0,0,0],
+                   [0,ry,0,0,0,0],
+                   [0,0,rz,0,0,0] ])
+    
 
     # Initial conditions for s0 and Sigma0
     # Compute the rest of sk using Eqs (2), (3), (4), and (5)
 
-    # ax.plot(x_coords, y_coords, z_coords,
-    #         '-r', label='Filtered trajectory')
+
+    ###### s0 is the same as in Part3
+    s0 = np.array([ [0,0,2,15,3.5,4.0] ]).T
+    Sig0 = np.identity(6)*0.01
+    
+    def predictS(s_k):
+        s_tilde = np.dot(A,s_k)+a
+        return s_tilde
+ 
+    def predictSig(Sig_k):
+        Sig_tilde = np.linalg.inv( np.dot( np.dot(A,Sig_k),(A.T) ) +
+                                   np.dot(B,(B.T)) )
+        return Sig_tilde
+
+    def updateSig(Sig_tilde):
+        Sig_k1 = np.linalg.inv( Sig_tilde + np.dot((C.T),C) )
+        return Sig_k1
+
+    def updateS(Sig_k1,Sig_tilde,s_tilde,m_k1):
+        s_k1 = np.dot( Sig_k1 , np.dot(Sig_tilde,s_tilde) +
+                                (np.matrix( np.dot((C.T),m_k1) )).T 
+                     )
+        return s_k1
+
+
+
+
+    myS = s0
+    sk  = s0
+    Sigk= Sig0
+	
+    for i in range(1,K):
+
+	s_inter   = predictS(sk)
+	Sig_inter = predictSig(Sigk)
+	Sigk      = updateSig(Sig_inter)
+	sk        = updateS(Sigk,Sig_inter,s_inter,m_trans[:,i])
+	myS = np.c_[myS, sk]
+    
+    myS = np.squeeze(np.asarray(myS))
```

Comment:
I was looking for a way to this this exact thing! Cool. I'll keep np.squeeze in mind.